pkg/test/e2e/cilium/cilium_test.go

```diff
@@ -0,0 +1,660 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cilium_test
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/observer"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	ciliumclientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned"
+	"google.golang.org/grpc"
+
+	"k8c.io/kubermatic/v2/pkg/resources"
+	"k8c.io/kubermatic/v2/pkg/test/e2e/utils"
+	"k8c.io/kubermatic/v2/pkg/test/e2e/utils/apiclient/models"
+	yamlutil "k8c.io/kubermatic/v2/pkg/util/yaml"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	kjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/util/rand"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var (
+	userconfig      string
+	accessKeyID     string
+	secretAccessKey string
+)
+
+const (
+	seed            = "kubermatic"
+	projectName     = "cilium-test-project"
+	userclusterName = "cilium-test-usercluster"
+	ciliumTestNs    = "cilium-test"
+)
+
+func init() {
+	flag.StringVar(&userconfig, "userconfig", "", "path to kubeconfig of usercluster")
+}
+
+func TestReadyCluster(t *testing.T) {
+	if userconfig == "" {
+		t.Logf("kubeconfig for usercluster not provided, test passes vacuously.")
+		t.Logf("to run against ready usercluster use following command")
+		t.Logf("go test ./pkg/test/e2e/cilium -v -race -tags e2e -timeout 30m -run TestReadyCluster -args --userconfig <USERCLUSTER KUBECONFIG>")
+		return
+	}
+	testUserCluster(t, userconfig)
+}
+
+func TestCiliumClusters(t *testing.T) {
+	accessKeyID = os.Getenv("AWS_ACCESS_KEY_ID")
+	if accessKeyID == "" {
+		t.Fatalf("AWS_ACCESS_KEY_ID not set")
+	}
+
+	secretAccessKey = os.Getenv("AWS_SECRET_ACCESS_KEY")
+	if secretAccessKey == "" {
+		t.Fatalf("AWS_SECRET_ACCESS_KEY not set")
+	}
+
+	tests := []struct {
+		name      string
+		proxyMode string
+	}{
+		{
+			name:      "ebpf proxy mode test",
+			proxyMode: resources.EBPFProxyMode,
+		},
+		{
+			name:      "ipvs proxy mode test",
+			proxyMode: resources.IPVSProxyMode,
+		},
+		{
+			name:      "iptables proxy mode test",
+			proxyMode: resources.IPTablesProxyMode,
+		},
+	}
+
+	var mu sync.Mutex
+
+	for _, test := range tests {
+		proxyMode := test.proxyMode
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			mu.Lock()
```

Comment:
not sure if the mutex is needed. There should not be issues if we create/delete multiple clusters at the same time. Or am I missing something?

Reply:
Without it sometimes cluster creation fails with 

> [POST /api/v1/projects][409] createProjectConflict

https://kubermatic.slack.com/archives/C01FBQ53EN6/p1644331513565219

Reply:
I hadn't enabled kubermatic-api logs when it happened. So not sure why it fails.

Reply:
I see what it is, but currently no idea why it happens. Its looks like a name conflict, but that shouldnt happen as we randomize the names